core/rpc_client.py

```diff
@@ -0,0 +1,230 @@
+             })
+
+        return self._socket.recv_json(), id
+
+
+class JSONRPCObjectProxy(object):
+    def __init__(self, connection, members, prefix=''):
+        """
+        This class serves as a base class for dynamically created classes on the
+        client side that represent server-side objects. Upon initialization,
+        this class takes the supplied methods and installs appropriate proxy methods
+        or properties into the object and class respectively. Because of that
+        class manipulation, this class must never be used directly.
+        Instead, it should be used as a base-class for dynamically created types
+        that mirror types on the server, like this:
+
+            proxy = type('SomeClassName', (ZMQJSONRPCObjectProxy, ), {})(socket, methods, prefix)
+
+        There is however, the class RemoteObjectCollection, which automates all that
+        and provides objects that are ready to use.
+
+        Exceptions on the server are propagated to the client. If the exception is not part
+        of the exceptions-module, a ZMQJSONRPCServerSideException is raised instead, which
+        contains information about the server side exception.
+
+        All RPC method names are prefixed with the supplied prefix, which is usually the
+        object name on the server plus a dot.
+
+        :param connection: ZMQREQConnection-object for remote calls.
+        :param members: List of strings to generate methods and properties.
+        :param prefix: Usually object name on the server plus dot.
+        """
+        self._properties = set()
+
+        self._connection = connection
+        self._prefix = prefix
+        self._add_member_proxies(members)
+
+    def _make_request(self, method, *args):
+        """
+        This method performs a JSON-RPC request via the object's ZMQ socket. If successful,
+        the result is returned, otherwise exceptions are raised. Server side exceptions are
+        raised using the same type as on the server if they are part of the exceptions-module.
+        Otherwise, a ZMQJSONRPCServerSideException is raised.
+
+        :param method: Method of the object to call on the remote.
+        :param args: Positional arguments to the method call.
+        :return: Result of the remote call if successful.
+        """
+        response, id = self._connection.json_rpc(self._prefix + method, *args)
+
+        if 'result' in response:
+            return response['result']
+
+        if 'error' in response:
+            if 'data' in response['error']:
+                exception_type = response['error']['data']['type']
+                exception_message = response['error']['data']['message']
+
+                if not hasattr(exceptions, exception_type):
+                    raise JSONRPCServerSideException(exception_type, exception_message)
+                else:
+                    exception = getattr(exceptions, exception_type)
+                    raise exception(exception_message)
```

Comment:
Yes, `__builtins__` is always available (without importing anything) and works exactly the same in both Python 2 and 3.

The one catch with it is that it works differently in the main (executed) module than in all other (imported) modules. In the main module it is an alias for the `__builtin__` module (in Python 2) or the `builtins` module (in Python 3). In all other modules, it is an alias for `__builtin__.__dict__` or `builtins.__dict__`, respectively. Odd behaviour, but consistent across all Python versions.

I'm happy leaving it as is... the only reason it caught my eye was that PyCharm flags `import builtins as exceptions` as an error.